src/compiler/moduleNameResolver.ts

```diff
@@ -816,6 +818,34 @@ namespace ts {
         return loadNodeModuleFromDirectory(extensions, candidate, failedLookupLocations, onlyRecordFailures, state, considerPackageJson);
     }
 
+    function parseNodeModuleFromPath(path: string): { packageDirectory: string, subModuleName: string } | undefined {
+        path = normalizePath(path);
+        const nodeModules = "/node_modules/";
+        const idx = path.indexOf(nodeModules);
+        if (idx === -1) {
+            return undefined;
+        }
+
+        const indexAfterNodeModules = idx + nodeModules.length;
+        let indexAfterPackageName = moveToNextDirectorySeparatorIfAvailable(path, indexAfterNodeModules);
+        if (path.charCodeAt(indexAfterNodeModules) === CharacterCodes.at) {
+            indexAfterPackageName = moveToNextDirectorySeparatorIfAvailable(path, indexAfterPackageName);

```

Comment:
A path should be a complete file path like `/node_modules/@asdf/pkg/index.d.ts`, not just `/node_modules/@asdf`.

Reply:
Are we concerned with malformed paths and packages? As a scenario, do we care if someone is mucking about and makes a malformed entry manually? I personally think tsserver shouldn't crash, and ideally we would surface a message to a user.

Do we have some way of handling situations like this gracefully?

Reply:
> I personally think tsserver shouldn't crash

That's just, like, your opinion, man. 😜 

But we shouldn't get a malformed path here because this is only called after a successful module resolution. And even if we did, I don't see a place where this function would crash.